### ISSUE TITLE
clippy thinks a tuple enum variant is a function call, leading it to think a closure is redundant

### ISSUE THREAD

**Issue by akabinds** [rust-lang/rust-clippy#9285]
### Summary

While calling `map_err` on the `io::Result` returned by `fs::read_to_string`, I map it to an `IoError` variant provided by my custom error type. This is what the error type looks like at the moment:

```rs
#[non_exhaustive]
#[derive(Debug, thiserror::Error)]
pub enum SectorError {
    #[error("failed to parse at line {1} on column {2}: {0:?}")]
    ParseError(&'static str, usize, usize),
    #[error("cache error: {0:?}")]
    CacheError(&'static str),
    #[error(transparent)]
    IoError(#[from] std::io::Error),
    #[error(transparent)]
    AllocError(#[from] std::alloc::AllocError),
}
```

As you can see, `IoError` is an enum variant, not a function like the Clippy lint warning suggests. I am of course able to silence the Clippy lint warning using an outer attribute to `allow` `clippy::redundant_closure`, but this should not even have to be the case. Clippy should be able to recognize that `IoError` is an enum variant, not a function call. If code written by users complies to the naming convention in Rust, this should be the conclusion as `struct`/`enum` names and `enum` variants are some of the items written using `CamelCase`. In the end, this should be the only thing it could be because from my current knowledge, other items whose naming convention is `CamelCase` can not be "called" directly like that. This includes *types*, *traits*, and *type parameters*.

### Lint Name

`redundant_closure`

### Reproducer

I tried this code:

```rust
pub fn mount_parser(path_to_source: &PathBuf) -> SectorResult<()> {
    let source = fs::read_to_string(path_to_source).map_err(|e| SectorError::IoError(e))?;

    let mut parser = Parser::new(Lexer::new(&source).lex()?);
    let query = parser.parse();

    if query.is_ok() {
        println!("{:?}", query.as_ref().unwrap());
    } else {
        eprintln!("{}", query.as_ref().unwrap_err());
    }

    Ok(())
}
```

I saw this happen:

```
  --> sector-core/native/sector_core/src/sec/mod.rs:14:61
   |
14 |     let source = fs::read_to_string(path_to_source).map_err(|e| SectorError::IoError(e))?;
   |                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `SectorError::IoError`
   |
   = note: `#[warn(clippy::redundant_closure)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

I expected to see this happen:

I expected to see no warnings.

### Version

```text
rustc 1.64.0-nightly (62b272d25 2022-07-21)
binary: rustc
commit-hash: 62b272d25c5bb8b6bb8ac73797d82b8b9a1eabda
commit-date: 2022-07-21
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by Jarcho** [comment]
Tuple structs and enum variants are also functions and can be used as suggested. The message could be reworded to say constructor instead of function.

**Comment by akabinds** [comment]
> Tuple structs and enum variants are also functions and can be used as suggested. The message could be reworded to say constructor instead of function.

I was not aware that tuple structs and enum variants were identified/classified as functions. In that case, is it impossible to alleviate this problem without explicitly using `#[allow(clippy::redundant_closure)]`?

**Comment by Jarcho** [comment]
Is there a reason to not use `.map_err(SectorError::IoError)`? The behaviour of the lint is unlikely to be changed.

**Comment by akabinds** [comment]
> Is there a reason to not use `.map_err(SectorError::IoError)`? The behaviour of the lint is unlikely to be changed.

Seems like that did the trick. I could have sworn I tried that before and it didn't work. Thank you for your time.
